### PR TITLE
RELEASES: Add Git Tag Job

### PR DESCRIPTION
**Pull request by glhays**
Closes: #58

AddTagJob

### CODE REVIEW

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Can you run the infrastructure build project and check in the newly generated dotnet.yaml

File: ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using System.Collections.Generic;
+using ADotNet.Models.Pipelines.GithubPipelines.DotNets.Tasks;
+using YamlDotNet.Serialization;
+
+namespace ADotNet.Models.Pipelines.GithubPipelines.DotNets
+{
+    public partial class TagJob
+    {
+        [YamlMember(Alias = "if")]
+        public string TagTrait = "github.event_name == 'pull_request' && github.event.action == 'closed'" +
```
Comment (cjdutoit) [inline ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs:16]: This did not work for me.  Can you provide some screenshots to show this is working only at the times you need it.

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Can you add some screenshots to show the tags were added as expected?

**Comment by glhays** [comment]
> Can you run the infrastructure build project and check in the newly generated dotnet.yaml

Infrastructure.Build run results....

Screenshot 2023-05-12 164332

**Comment by glhays** [comment]
> Can you add some screenshots to show the tags were added as expected?

Only added in the the Test.Console

File: ADotNet/Models/Pipelines/GithubPipelines/DotNets/Tasks/TagTask.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// ---------------------------------------------------------------------------
+
+using YamlDotNet.Serialization;
+
+namespace ADotNet.Models.Pipelines.GithubPipelines.DotNets.Tasks
+{
+    public partial class TagTask : GithubTask
+    {
+        [YamlMember(Order = 1, ScalarStyle = YamlDotNet.Core.ScalarStyle.Literal)]
+        public string Run = "git tag -a \"release-${{ github.run_number }}\" -m \"Release ${{ github.run_number }}\"\r\ngit push origin --tags";
```
Comment (cjdutoit) [inline ADotNet/Models/Pipelines/GithubPipelines/DotNets/Tasks/TagTask.cs:14]: Can we change this so it uses the version number instead.  I can provide code from my POC.

Comment (glhays) [reply ADotNet/Models/Pipelines/GithubPipelines/DotNets/Tasks/TagTask.cs:14]: Sure

**Comment by cjdutoit** [comment]
@glhays here is a working version of my POC.   I could not get a combined version of tagging as well as the other jobs to work together in a single yaml file as you need `types: closed` for the merge event.   In the same file that job was always skipped.
```
on:
  pull_request:
    types:
      - closed
```
I created a seperate yaml file just for tagging and it seems to do the job.  It pulls the version from the proj file and use the version to create a tag `Release-1.0.0`

Also, I could not add a tag without authenticating.  I had to create a Personal Access Token.   I did not want to hardcode my name on the user.name OR user.email, but adding the fake details like this still works with the access token.

```yaml
Here is my current version of the tagging
```yaml
name: Add Git Release Tag on Merge to Main
on:
  pull_request:
    types:
      - closed
    branches:
      - main

jobs:
  add_release_tag:
    if: github.event.pull_request.merged && github.event.pull_request.base.ref == 'main' && startsWith(github.event.pull_request.title, 'RELEASES:') && (contains(github.event.pull_request.labels.*.name, 'RELEASE') || contains(github.event.pull_request.labels.*.name, 'RELEASE'))  

    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Extract Version Number
        id: extract_version
        run: echo "::set-output name=version_number::$(grep -oP '(?<=<Version>)[^<]+' BuildTestApp/BuildTestApp.csproj)"

      - name: Print Version Number
        run: echo "Version number is ${{ steps.extract_version.outputs.version_number }}"

      - name: Configure Git
        run: |
          git config user.name "Add Git Release Tag Action"
          git config user.email "github.action@noreply.github.com"

      - name: Authenticate with GitHub
        uses: actions/checkout@v2
        with:
          token: ${{ secrets.PAT_FOR_TAGGING }}

      - name: Add Git Tag - Release
        run: |
          git tag -a "release-${{ steps.extract_version.outputs.version_number }}" -m "Release ${{ steps.extract_version.outputs.version_number }}"
          git push origin --tags
```

**Comment by cjdutoit** [comment]
> > Can you run the infrastructure build project and check in the newly generated dotnet.yaml
> 
> Infrastructure.Build run results....
> 
> Screenshot 2023-05-12 164332

#55 will add the tag for the push filter. 
```
on:
  push:
    tags:
    - RELEASE
```

**Comment by cjdutoit** [comment]
> > Can you add some screenshots to show the tags were added as expected?
> 
> Only added in the the Test.Console

You will need to provide more evidence.  Take the generated yaml and add it to a private repo to test the outcomes. i.e. the below one showed the add_tag was skipped as we dont do it on a PR.
image

but on this one, you can see that the tagging happened on main.

image
image

**Comment by glhays** [comment]
I think it best I revert my code and close the pr, as the ASk for this issue has changed. Understandably so due to your testing result(s). Any thoughts?

File: ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                " && github.event.pull_request.merged && startsWith(github.event.pull_request.title, 'RELEASES:')" + "" +
+                    " && contains(github.event.pull_request.labels.*.name, 'RELEASES')";
+
+        [YamlMember(Alias = "needs")]
+        public string[] Needs = { "build", "test" };
```
Comment (cjdutoit) [inline ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs:21]: I dont think we should hardcode the values "build" & "test".  This libary should give the consumer the flexibility to build it up as needed.

```suggestion
        [YamlMember(Alias = "needs", DefaultValuesHandling = DefaultValuesHandling.OmitDefaults)]
        public string[] Needs = { get; set; };
```

File: ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs (hunk 1 above)
Comment (cjdutoit) [reply ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs:16]: we should also not hardcode this.  consumer should be able to build this section out themselves.

Comment (glhays) [reply ADotNet/Models/Pipelines/GithubPipelines/DotNets/TagJob.cs]: Will try to find some time this weekend to get a test repo setup as suggested.

**Comment by cjdutoit** [comment]
> I think it best I revert my code and close the pr, as the ASk for this issue has changed. Understandably so due to your testing result(s). Any thoughts?

Thanks.   I think if you can do just the Nuget Deploy as a working POC as per #50, then we can just add the `if:` condition to that step one this is done.  

It would probably not take that long now to finish off #58 from what we have done on the POC.

**Pull request closed without merging**